In [1]:
from sys import prefix

from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA


In [2]:
data=pd.read_csv('C:/Users\ESHITA\Downloads/accepted_2007_to_2018Q4.csv/accepted_2007_to_2018Q4.csv')
print(data.head())

C:\Users\ESHITA\AppData\Local\Temp\ipykernel_35928\2986039858.py:1: DtypeWarning: Columns (0,19,49,59,118,129,130,131,134,135,136,139,145,146,147) have mixed types. Specify dtype option on import or set low_memory=False.
  data=pd.read_csv('C:/Users\ESHITA\Downloads/accepted_2007_to_2018Q4.csv/accepted_2007_to_2018Q4.csv')


         id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  68407277        NaN     3600.0       3600.0           3600.0   36 months   
1  68355089        NaN    24700.0      24700.0          24700.0   36 months   
2  68341763        NaN    20000.0      20000.0          20000.0   60 months   
3  66310712        NaN    35000.0      35000.0          35000.0   60 months   
4  68476807        NaN    10400.0      10400.0          10400.0   60 months   

   int_rate  installment grade sub_grade  ... hardship_payoff_balance_amount  \
0     13.99       123.03     C        C4  ...                            NaN   
1     11.99       820.28     C        C1  ...                            NaN   
2     10.78       432.66     B        B4  ...                            NaN   
3     14.85       829.90     C        C5  ...                            NaN   
4     22.45       289.91     F        F1  ...                            NaN   

  hardship_last_payment_amount disbursement_

In [3]:
N_data=data[["loan_amnt","term","int_rate", "installment","grade","sub_grade","fico_range_low","annual_inc","dti","emp_length","purpose","open_acc","total_acc","revol_util","delinq_2yrs",
"inq_last_6mths","loan_status"]]
print(N_data.head())

   loan_amnt        term  int_rate  installment grade sub_grade  \
0     3600.0   36 months     13.99       123.03     C        C4   
1    24700.0   36 months     11.99       820.28     C        C1   
2    20000.0   60 months     10.78       432.66     B        B4   
3    35000.0   60 months     14.85       829.90     C        C5   
4    10400.0   60 months     22.45       289.91     F        F1   

   fico_range_low  annual_inc    dti emp_length             purpose  open_acc  \
0           675.0     55000.0   5.91  10+ years  debt_consolidation       7.0   
1           715.0     65000.0  16.06  10+ years      small_business      22.0   
2           695.0     63000.0  10.78  10+ years    home_improvement       6.0   
3           785.0    110000.0  17.06  10+ years  debt_consolidation      13.0   
4           695.0    104433.0  25.37    3 years      major_purchase      12.0   

   total_acc  revol_util  delinq_2yrs  inq_last_6mths loan_status  
0       13.0        29.7          0.0     

In [4]:
N_data["term"].dtype

dtype('O')

In [5]:
d1=pd.get_dummies(N_data,columns=['term','grade','sub_grade','emp_length','purpose'],prefix={'term':'term','grade':'grade','sub_grade':'sub_grade','emp_length':'emp_length','purpose':'purpose'},dtype=int)

In [6]:
d1.head()

loan_amnt  int_rate  installment  fico_range_low  annual_inc    dti  \
0     3600.0     13.99       123.03           675.0     55000.0   5.91   
1    24700.0     11.99       820.28           715.0     65000.0  16.06   
2    20000.0     10.78       432.66           695.0     63000.0  10.78   
3    35000.0     14.85       829.90           785.0    110000.0  17.06   
4    10400.0     22.45       289.91           695.0    104433.0  25.37   

   open_acc  total_acc  revol_util  delinq_2yrs  ...  \
0       7.0       13.0        29.7          0.0  ...   
1      22.0       38.0        19.2          1.0  ...   
2       6.0       18.0        56.2          0.0  ...   
3      13.0       17.0        11.6          0.0  ...   
4      12.0       35.0        64.5          1.0  ...   

   purpose_home_improvement purpose_house  purpose_major_purchase  \
0                         0             0                       0   
1                         0             0                       0   
2                         1             0                       0   
3                         0             0                       0   
4                         0             0                       1   

   purpose_medical  purpose_moving  purpose_other  purpose_renewable_energy  \
0                0               0              0                         0   
1                0               0              0                         0   
2                0               0              0                         0   
3                0               0              0                         0   
4                0               0              0                         0   

   purpose_small_business  purpose_vacation  purpose_wedding  
0                       0                 0                0  
1                       1                 0                0  
2                       0                 0                0  
3                       0                 0                0  
4                       0                 0                0  

[5 rows x 81 columns]

In [7]:
y=((d1['loan_status']=="Fully Paid").astype(int))
y.head()

0    1
1    1
2    1
3    0
4    1
Name: loan_status, dtype: int32

In [8]:
X=d1.drop('loan_status',axis=1)
X.head()

loan_amnt  int_rate  installment  fico_range_low  annual_inc    dti  \
0     3600.0     13.99       123.03           675.0     55000.0   5.91   
1    24700.0     11.99       820.28           715.0     65000.0  16.06   
2    20000.0     10.78       432.66           695.0     63000.0  10.78   
3    35000.0     14.85       829.90           785.0    110000.0  17.06   
4    10400.0     22.45       289.91           695.0    104433.0  25.37   

   open_acc  total_acc  revol_util  delinq_2yrs  ...  \
0       7.0       13.0        29.7          0.0  ...   
1      22.0       38.0        19.2          1.0  ...   
2       6.0       18.0        56.2          0.0  ...   
3      13.0       17.0        11.6          0.0  ...   
4      12.0       35.0        64.5          1.0  ...   

   purpose_home_improvement  purpose_house  purpose_major_purchase  \
0                         0              0                       0   
1                         0              0                       0   
2                         1              0                       0   
3                         0              0                       0   
4                         0              0                       1   

   purpose_medical  purpose_moving  purpose_other  purpose_renewable_energy  \
0                0               0              0                         0   
1                0               0              0                         0   
2                0               0              0                         0   
3                0               0              0                         0   
4                0               0              0                         0   

   purpose_small_business  purpose_vacation  purpose_wedding  
0                       0                 0                0  
1                       1                 0                0  
2                       0                 0                0  
3                       0                 0                0  
4                       0                 0                0  

[5 rows x 80 columns]

In [9]:
X['revol_util'].dtype

dtype('float64')

In [10]:
X.isnull().sum().sort_values(ascending=False)


revol_util         1835
dti                1744
inq_last_6mths       63
delinq_2yrs          62
open_acc             62
                   ... 
sub_grade_D4          0
sub_grade_D5          0
sub_grade_E2          0
sub_grade_E3          0
purpose_wedding       0
Length: 80, dtype: int64

In [11]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [12]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler()),
    ('model', LogisticRegression(max_iter=1000))
])


In [13]:
pipeline.fit(X_train,y_train)

Pipeline(steps=[('imputer', SimpleImputer(strategy='median')),
                ('scaler', StandardScaler()),
                ('model', LogisticRegression(max_iter=1000))])

In [14]:
pipeline.score(X_test,y_test)

0.6276979968638102

In [15]:
y.value_counts(normalize=True)


loan_status
0    0.523709
1    0.476291
Name: proportion, dtype: float64

In [16]:
from sklearn.metrics import classification_report, roc_auc_score

y_pred = pipeline.predict(X_test)
y_proba = pipeline.predict_proba(X_test)[:, 1]

print(classification_report(y_test, y_pred))
print("ROC-AUC:", roc_auc_score(y_test, y_proba))


              precision    recall  f1-score   support

           0       0.64      0.65      0.65    237500
           1       0.61      0.60      0.61    214641

    accuracy                           0.63    452141
   macro avg       0.63      0.63      0.63    452141
weighted avg       0.63      0.63      0.63    452141

ROC-AUC: 0.6741030339668759


When Logistic Regression gives ~0.62–0.64 accuracy and ~0.67 AUC on a balanced dataset, it usually means the problem contains non-linear interactions.